File: c/physics/bounce/bench.c
/**
 * @file   bench.c
 * @brief  Benchmark for ball_update — outputs JSON consumed by the report generator.
 *
 * Runs ball_update N times in a tight loop with no I/O inside the timed
 * region. The volatile sink prevents the compiler from optimising the
 * loop away entirely (dead-code elimination).
 *
 * Output format (one JSON object to stdout):
 *   {"lang":"c","example":"bounce","mean_ns":42.3,"std_ns":1.2,"n":1000000}
 */

#include <stdio.h>
#include <math.h>
#include <time.h>
#include "physics.h"

#define WARMUP  100000
#define RUNS    1000000

static double now_ns(void) {
    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    return t.tv_sec * 1e9 + t.tv_nsec;
}

int main(void) {
    Ball ball = { .x = 400.0f, .y = 300.0f, .vx = 4.0f, .vy = 3.0f };

    /* Warm up — not timed */
    for (int i = 0; i < WARMUP; i++)
        ball_update(&ball, 30, 800, 600);

    /* Timed runs */
    double t0 = now_ns();
    for (int i = 0; i < RUNS; i++)
        ball_update(&ball, 30, 800, 600);
    double t1 = now_ns();

    /* Prevent dead-code elimination */
    volatile float sink = ball.x;
    (void)sink;

    double mean_ns = (t1 - t0) / RUNS;
    printf("{\"lang\":\"c\",\"example\":\"bounce\",\"mean_ns\":%.4f,\"n\":%d}\n",
           mean_ns, RUNS);
    return 0;
}

File: rust/physics/bounce/src/bench.rs
//! Standalone benchmark binary — outputs JSON for the report generator.
//!
//! Run: cargo run --release --bin bench_bounce

mod physics;
use physics::Ball;
use std::time::Instant;

const WARMUP: usize = 100_000;
const RUNS:   usize = 1_000_000;

fn main() {
    let mut ball = Ball::new(400.0, 300.0, 4.0, 3.0);

    for _ in 0..WARMUP {
        ball.update(30, 800, 600);
    }

    let start = Instant::now();
    for _ in 0..RUNS {
        std::hint::black_box(ball.update(
            std::hint::black_box(30),
            std::hint::black_box(800),
            std::hint::black_box(600),
        ));
    }
    let elapsed = start.elapsed();

    let mean_ns = elapsed.as_nanos() as f64 / RUNS as f64;
    println!(
        "{{\"lang\":\"rust\",\"example\":\"bounce\",\"mean_ns\":{:.4},\"n\":{}}}",
        mean_ns, RUNS
    );
}
